### ISSUE TITLE
[Request for Integration of Web-based GO IDE]

### ISSUE THREAD

**Issue by Secarian** [tobychui/arozos#116]
**Problem description:**
The current process for writing and testing small code snippets within the ArozOS environment is not user-friendly.

**Desired solution:**
I request the integration of a web-based IDE for GO, such as [Wide](https://github.com/cnsuhao/wide), as a standard package in the ArozOS project. This integration would allow users to easily write and test GO applications within the ArozOS interface, making the development process of new apps simpler and more accessible, especially for beginners. The ability to export finished installation packages directly from the IDE would further streamline the development process.

**Considered alternatives:**
No alternatives have been considered.

**Additional context:**
Currently, there is no simple way to create apps and installation packages in the ArozOS environment. The integration of a standard IDE, such as [Wide](https://github.com/cnsuhao/wide), would provide this functionality and result in a significant improvement in user-friendliness. Furthermore, it could allow for the integration of a custom installation package format, enabling the creation, detection, and handling of installations using a custom file extension recognized by ArozOS.

**Comment by markusbkk** [comment]
I was tinkering with that idea a while back.
Personally, I think something like VSCode would be a good option too. Since it's already widely adopted, has a decent plugin architecture, support for many programming languages and is somewhat straightforward to embed.

Don't get me wrong. Wide looks nice, too (and some of the code might help with embedding any other editor into ArozOS. I'm thinking a generalized API) but it seems focused on Go development (not necessarily the target audience for those using ArozOS as clients) and isn't nearly as battle tested as VSCode.

**Comment by yeungalan** [comment]
Hi! do you guys tried code studio? This web app basically behaves like VSCode and probably fits your need
![image](https://user-images.githubusercontent.com/12872405/216573518-799e4477-a537-4b54-9635-9c46d4023b41.png)

**Comment by markusbkk** [comment]
> Hi! do you guys tried code studio in arozos? This web app behaves like VSCode and probably fits your need
> 
> Here is the Code Studio IDE screenshot :) ![image](https://user-images.githubusercontent.com/12872405/216573518-799e4477-a537-4b54-9635-9c46d4023b41.png)

Yea. I saw that. I had some issues opening and saving files, AFAIR.
I think this is also not based on the current VSCode embedding strategy, although I might be wrong.

**Comment by tobychui** [comment]
Yes, this is technically only the Monaco Editor part of the VScode but not the whole eco-system that most of the VScode user used to.  

And for @markusbkk 's issue, some minor functions of this editor is still work in progress and still waiting for the go compiler integration. My goal was to provide most basic Go IDE support so developers can develop and compile ArozOS inside ArozOS.

**Comment by markusbkk** [comment]
> Yes, this is technically only the Monaco Editor part of the VScode but not the whole eco-system that most of the VScode user used to.
> 
> And for @markusbkk 's issue, some minor functions of this editor is still work in progress and still waiting for the go compiler integration. My goal was to provide most basic Go IDE support so developers can develop and compile ArozOS inside ArozOS.

Thought so. There's actually a more complete VSCode integration that also supports some of the plugins and themes.

**Comment by markusbkk** [comment]
@tobychui 

AFAIK, this one is more "proper" now. It's similar to how Microsoft implements their own online code editor on GitHub.

https://github.com/Felx-B/vscode-web

**Comment by tobychui** [comment]
@markusbkk 

Integrating another web-based webapps not something hard to do. The current issue is how to integrate it while keeping file system sandbox and user isolation in place, with expected supports for future cluster architecture expansion. 

It is relatively easy to integrate it (vscode-web from Felx-B) into your instance of ArozOS if you are using it in a one-person per host settings. However, if you are hosting it for multiple users with different storage pool configurations (which I design this system for this particular purpose), it is really hard to perform a proper & flexible integration without an expert who is experienced with container technology contributing to this project. 

Though, I personally love the idea to run 3rd party apps without any glue-code and modification to the original app, "subservice apps" are currently the closest we have.

**Comment by markusbkk** [comment]
> @markusbkk
> 
> Integrating another web-based webapps not something hard to do. The current issue is how to integrate it while keeping file system sandbox and user isolation in place, with expected supports for future cluster architecture expansion.
> 
> It is relatively easy to integrate it (vscode-web from Felx-B) into your instance of ArozOS if you are using it in a one-person per host settings. However, if you are hosting it for multiple users with different storage pool configurations (which I design this system for this particular purpose), it is really hard to perform a proper & flexible integration without an expert who is experienced with container technology contributing to this project.
> 
> Though, I personally love the idea to run 3rd party apps without any glue-code and modification to the original app, "subservice apps" are currently the closest we have.

Oh. I absolutely understand this. I was talking solely from the JS frontend side. Obviously, things like the FS driver still need to be specialized.
